server/src/main/java/io/druid/server/namespace/URIExtractionNamespaceFunctionFactory.java

```diff
@@ -0,0 +1,183 @@
+import com.metamx.common.logger.Logger;
+import io.druid.common.utils.JodaUtils;
+import io.druid.data.input.MapPopulator;
+import io.druid.query.extraction.namespace.ExtractionNamespaceFunctionFactory;
+import io.druid.query.extraction.namespace.JDBCExtractionNamespace;
+import io.druid.query.extraction.namespace.URIExtractionNamespace;
+import io.druid.segment.loading.DataSegmentPuller;
+import io.druid.segment.loading.URIDataPuller;
+import io.druid.server.namespace.cache.NamespaceExtractionCacheManager;
+import org.joda.time.format.DateTimeFormatter;
+import org.joda.time.format.ISODateTimeFormat;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentMap;
+import java.util.regex.Pattern;
+
+/**
+ *
+ */
+public class URIExtractionNamespaceFunctionFactory implements ExtractionNamespaceFunctionFactory<URIExtractionNamespace>
+{
+  private static final Logger log = new Logger(URIExtractionNamespaceFunctionFactory.class);
+  private final Map<String, DataSegmentPuller> pullers;
+
+  @Inject
+  public URIExtractionNamespaceFunctionFactory(
+      Map<String, DataSegmentPuller> pullers
+  )
+  {
+    this.pullers = pullers;
+  }
+
+  @Override
+  public Function<String, String> build(URIExtractionNamespace extractionNamespace, final Map<String, String> cache)
+  {
+    return new Function<String, String>()
+    {
+      @Nullable
+      @Override
+      public String apply(String input)
+      {
+        if (Strings.isNullOrEmpty(input)) {
+          return null;
+        }
+        return Strings.emptyToNull(cache.get(input));
+      }
+    };
+  }
+
+  @Override
+  public Callable<String> getCachePopulator(
+      final URIExtractionNamespace extractionNamespace,
+      final String lastVersion,
+      final Map<String, String> cache
+  )
+  {
+    final long lastCached = lastVersion == null ? JodaUtils.MIN_INSTANT : Long.parseLong(lastVersion);
+    return new Callable<String>()
+    {
+      @Override
+      public String call()
+      {
+        final URI originalUri = extractionNamespace.getUri();
+        final DataSegmentPuller pullerRaw = pullers.get(originalUri.getScheme());
+        if (pullerRaw == null) {
+          throw new IAE(
+              "Unknown loader type[%s].  Known types are %s",
+              originalUri.getScheme(),
+              pullers.keySet()
+          );
+        }
+        if (!(pullerRaw instanceof URIDataPuller)) {
+          throw new IAE(
+              "Cannot load data from location [%s]. Data pulling from [%s] not supported",
+              originalUri.toString(),
+              originalUri.getScheme()
+          );
+        }
+        final URIDataPuller puller = (URIDataPuller) pullerRaw;
+        final String versionRegex = extractionNamespace.getVersionRegex();
+        final URI uri = puller.getLatestVersion(
```

Comment:
I didn't migrate over the notification in the PR master comment properly, will update

https://github.com/druid-io/druid/pull/1443 is the broken out part of this PR